server/src/test/java/org/elasticsearch/index/mapper/DocumentParserTests.java

```diff
@@ -2574,6 +2578,375 @@ public void testDeeplyNestedDocument() throws Exception {
+        DocumentMapper mapper = createDocumentMapper(mapping(b -> {
+            b.startObject("metrics");
+            {
+                b.field("type", "object").field("subobjects", false);
+                b.startObject("properties");
+                {
+                    b.startObject("service.time");
+                    b.field("type", "long");
+                    b.field("null_value", -1);
+                    b.endObject();
+                }
+                {
+                    b.startObject("service.time.max");
+                    b.field("type", "long");
+                    b.field("null_value", -1);
+                    b.endObject();
+                }
+                {
+                    b.startObject("service.time.min");
+                    b.field("type", "long");
+                    b.field("null_value", -1);
+                    b.endObject();
+                }
+                {
+                    b.startObject("service.time.avg");
+                    b.field("type", "long");
+                    b.field("null_value", -1);
+                    b.endObject();
+                }
+                b.endObject();
+            }
+            b.endObject();
+        }));
+
+        ParsedDocument parsedDocument = mapper.parse(source("""
+            {
+              "metrics": {
+                "service.time" : null,
+                "service": {
+                  "time" : {
+                    "min" : null
+                  },
+                  "time.max" : null
+                },
+                "service.time.avg" : null
+              }
+            }
+            """));
+
+        assertNotNull(parsedDocument.rootDoc().getField("metrics.service.time"));
+        assertNotNull(parsedDocument.rootDoc().getField("metrics.service.time.min"));
+        assertNotNull(parsedDocument.rootDoc().getField("metrics.service.time.max"));
+        assertNotNull(parsedDocument.rootDoc().getField("metrics.service.time.avg"));
+        assertNull(parsedDocument.dynamicMappingsUpdate());
+    }
+
+    public void testSubobjectsFalseDocsWithInnerObjectMappedAsNonObject() throws Exception {
+        DocumentMapper mapper = createDocumentMapper(mapping(b -> {
+            b.startObject("metrics");
+            {
+                b.field("type", "object").field("subobjects", false);
+                b.startObject("properties");
+                {
+                    b.startObject("service.time");
+                    b.field("type", "long");
+                    b.endObject();
+                }
+                b.endObject();
+            }
+            b.endObject();
+        }));
+        ParsedDocument parsedDocument = mapper.parse(source("""
+            {
+              "metrics": {
+                "service.time" : {
+                  "min" : 1
+                }
+              }
+            }
+            """));
```

Comment:
Does this test run? I remember the problem here being that we would provide the inner object to be parsed by the long `service.time` field, which would not work. That is because service.time is mapped, but not as an object. I am not completely following how you worked around that, I am also quite fuzzy on this because it's been a while since I worked on this issue.

Reply:
It did run because of a bug discovered thanks to your comment above. 
Ref:
> I would expect this to break things, as it effectively ignores the subobjects setting at each object level (it can be set within any object)

Reply:
Got it. This is the third problem we discussed. When you do have an object in the document, and the field is mapped but not as an object, there are cases in which you do need to provide the object as-is to the mapper for parsing (e.g. geo_point) and the majority of the cases in which you can just throw an error (e.g. because long does not support parsing objects). We are covered for the error situation, that happens already, but we need to cover the geo_point scenario, and we don't have a systematic way to say: "does the mapper support parsing objects?". We should avoid instanceof checks and rather introduce a way to discriminate between mappers that can and cannot parse objects natively.

Reply:
So, if I understood correctly, when `subobject: false` we want to be able to provide the entire object as-is to the mapper for parsing if either:

1. There is a dynamic template for such field or 
2. There is a mapped field of any type that can natively parse objects.

While in the case of this particular test, where there is a mapped field but the type cannot parse objects, we just want an exception.

Did I understood correctly or am I missing something?

Reply:
Sounds accurate!

Reply:
I tripped on the following sentence:

> While in the case of this particular test, where there is a mapped field but the type cannot parse objects, we just want an exception.

Not, we want to parse its sub-fields and ignore their object container. That is one of the challenges of this change. Sorry about the confusion.

Reply:
Got it! Thanks for clarifying. 👍